sql/core/src/main/java/org/apache/spark/sql/sources/v2/MicroBatchReadSupport.java

```diff
@@ -0,0 +1,22 @@
+package org.apache.spark.sql.sources.v2;
+
+import java.util.Optional;
+
+import org.apache.spark.annotation.InterfaceStability;
+import org.apache.spark.sql.sources.v2.reader.MicroBatchReader;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A mix-in interface for {@link DataSourceV2}. Data sources can implement this interface to
+ * provide streaming micro-batch data reading ability.
+ */
+@InterfaceStability.Evolving
+public interface MicroBatchReadSupport extends DataSourceV2 {
+  /**
+   * Creates a {@link MicroBatchReader} to scan a batch of data from this data source.
+   *
+   * @param options the options for the returned data source reader, which is an immutable
+   *                case-insensitive string-to-string map.
+   */
+  MicroBatchReader createMicroBatchReader(Optional<StructType> schema, String checkpointLocation, DataSourceV2Options options);
```

Comment:
Do we use the checkpoint location for anything other than recording the start position?  Maybe we should keep this regardless, I'm not sure.

Reply:
We do. In FileStreamSource, for instance, we keep a list of all files that have been seen.

Reply:
Ah, makes sense.  We should document what this is for (i.e. it is scratch space that the source can use if it needs to record anything for recovery).

Reply:
What about putting a best effort limit here?  I think it could be bytes or records.  Slight preference for records.

Reply:
How would users specify this parameter? If it's as an option(), it'll already be packed inside DataSourceV2Options.

If it's not as an option, will/can/should we add this best effort limit as a top-level concept to DataStreamReader?